src/stream.js

```diff
@@ -54,62 +64,97 @@ module.exports = ({ id, name, send, onEnd = () => {}, type = 'initiator', maxMsg
       stream.timeline.close = Date.now()
       onEnd(endErr)
     }
+    if (sinkClosedDefer) sinkClosedDefer.resolve()
+  }
+
+  const _send = (message) => {
+    if (!sinkEnded) {
+      send(message)
+    }
   }
 
+  /** @type {MuxedStream} */
   const stream = {
+    // Close for both Reading and Writing
+    close: () => Promise.all([
+      stream.closeRead(),
+      stream.closeWrite()
+    ]),
     // Close for reading
-    close: () => stream.source.end(),
+    closeRead: () => stream.source.end(),
+    // Close for writing
+    closeWrite: () => {
+      if (sinkEnded) {
+        return
+      }
+
+      if (sinkInProgress) {
+        sinkClosedDefer = pDefer()
+        writeCloseController.abort()
+        return sinkClosedDefer.promise
+      }
+
+      return stream.sink([])
```

Comment:
I think I'm starting to wrap my head around these streams, and I have one question regarding the current code in relation to the spec: does it make sense to create a new stream right before closing it here? Does the spec require this, or could we simply replace it with `onSinkEnd()` and prohibit a the `sink` function from being called if `sinkEnded`?

I don't see anything indicating that there needs to be a stream in both directions in https://github.com/libp2p/specs/tree/master/mplex#opening-a-new-stream.

Reply:
the advantage of doing so is that we inform the other party that we will not write to it anymore and only expect read and their side will also be close for reading